### PR TITLE
Handle sensitive information being inside a list in resource_dict

### PR DESCRIPTION
**Pull request by dbasunag**
##### Short description: 
Current code hash_resource_dict() does not provide a flexible way to hide fields like userData, for virtual machines (it could be present in spec.template.spec.volumes)

## Summary by CodeRabbit

- **New Features**
	- Introduced a new function to securely replace sensitive values in nested dictionaries.
	- Updated key representation format for resource management.
	- Added a new property to access user data in virtual machine configurations.
- **Bug Fixes**
	- Enhanced error handling for invalid inputs in the new function.
- **Tests**
	- Implemented a comprehensive test suite for the new sensitive value replacement function.
- **Chores**
	- Added a new testing environment configuration for unit tests.

### CODE REVIEW

**Comment by redhat-qe-bot2** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
    * You can add extra args to the Podman build command
        * Example: `/build-and-push-container --build-arg OPENSHIFT_PYTHON_WRAPPER_COMMIT=<commit_hash>`
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest python-module-install`: Retest python-module-install
 * `/retest all`: Retest all

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: ocp_resources/resource.py [hunk 1]
```diff
@@ -134,9 +133,27 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
     return None
 
 
-# Exceptions classes
+def change_dict_value_to_hashed(resource_dict: Dict[Any, Any], key_name: str) -> Dict[Any, Any]:
```
Comment (myakove) [inline ocp_resources/resource.py:136]: find a better name for it?

- `mask_key_values`
- `replace_key_with_hashed_value`
- `hide_key_in_dict`

Comment (myakove) [inline ocp_resources/resource.py:136]: Please add tests.

File: ocp_resources/resource.py [hunk 2]
```diff
@@ -134,9 +133,27 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+    Recursively search a nested dictionary for a given key and changes its value to "******" if found.
+
+    Args:
+        resource_dict: The nested dictionary to search.
+        key_name: The key to find.
 
-# End Exceptions classes
+    Returns:
+        The modified dictionary.
+    """
```
Comment (myakove) [inline ocp_resources/resource.py:146]: we need to check that arg is dict

```
    if not isinstance(resource_dict, dict):
        raise ValueError("Expected a dictionary as the first argument")
```

File: ocp_resources/resource.py [hunk 3]
```diff
@@ -134,9 +133,27 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+    if isinstance(resource_dict, dict):
+        for key, value in resource_dict.items():
+            if key == key_name:
+                resource_dict[key] = "******"
+            elif isinstance(value, dict):
+                resource_dict[key] = change_dict_value_to_hashed(value, key_name)
+            elif isinstance(value, list):
```
Comment (myakove) [inline ocp_resources/resource.py:153]:
```suggestion
                    elif isinstance(value, list) and any(isinstance(item, dict) for item in value):
```

to enter only if we have dict under the list

File: ocp_resources/sealed_secret.py [hunk 4]
```diff
@@ -50,4 +50,4 @@ def to_dict(self) -> None:
 
     @property
     def keys_to_hash(self):
-        return ["spec..data", "spec..encryptedData"]
+        return ["data", "encryptedData"]
```
Comment (myakove) [inline ocp_resources/sealed_secret.py:53]: This change the logic of the original usage
Now every `data` key will be hashed, not only the one under `spec`
we need to think if this is make sense or not.

File: ocp_resources/resource.py [hunk 5]
```diff
@@ -15,7 +15,6 @@
 import kubernetes
 from kubernetes.dynamic import DynamicClient, ResourceInstance
 import yaml
-from benedict import benedict
```
Comment (myakove) [inline ocp_resources/resource.py:18]: `benedict` is no longer used and should be removed from the project.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/sealed_secret.py (hunk 4 above)
Comment (rnetser) [reply ocp_resources/sealed_secret.py:53]: i think that we should be implicit here and keep the current logic

Comment (dbasunag) [reply ocp_resources/sealed_secret.py:53]: Do we have any instance of same key? As far as I have seen, sometimes a key in spec might be present in status as well (in this case both should be hidden) or in case of templates and resource using templates, duplicate keys might be present as well (this might be a case where I don't know if hiding all is smart or not). But I need both your guidance on this one a bit more. I am a bit blind about my own usecase and have a mindset of `show only what we must` if not specified.

File: ocp_resources/resource.py (hunk 3 above)
Comment (dbasunag) [reply ocp_resources/resource.py:153]: I don't think it would be needed. 
```
            elif isinstance(value, list):
                for key_list, value_list in enumerate(value):
                    value[key_list] = change_dict_value_to_hashed(value_list, key_name)
```
This call `change_dict_value_to_hashed(value_list, key_name)` on value will simply return the value_list, as value_list is a list. and change_dict_value_to_hashed() would only work on dicts.

But may be I should add a check for list in line 156? What do you think?

File: ocp_resources/resource.py (hunk 2 above)
Comment (dbasunag) [reply ocp_resources/resource.py:146]: Ok, will add. But should we talk on this first? I am not sure how to preserve what we have with this PR.

File: ocp_resources/resource.py (hunk 5 above)
Comment (dbasunag) [reply ocp_resources/resource.py:18]: ok. Will do.

File: ocp_resources/resource.py (hunk 3 above)
Comment (myakove) [reply ocp_resources/resource.py:153]: `But may be I should add a check for list in line 156? What do you think?` 
I don;t understand.

File: ocp_resources/sealed_secret.py (hunk 4 above)
Comment (myakove) [reply ocp_resources/sealed_secret.py:53]: it doesn't matter if we have the same key or not, maybe we do not have it now but what about later?

For example this is bad

```
In [11]: qq = {"a": ["1", "2"], "b": {"b1": [1,2]}}

In [12]: change_dict_value_to_hashed(qq, "b")
Out[12]: {'a': ['1', '2'], 'b': '******'}

In [13]: change_dict_value_to_hashed(qq, "a")
Out[13]: {'a': '******', 'b': '******'}

In [14]: change_dict_value_to_hashed(qq, "2")
Out[14]: {'a': '******', 'b': '******'}
```

We should be able to send `a.1` to mask only `1` under `a`

WDYT?

File: ocp_resources/resource.py (hunk 3 above)
Comment (dbasunag) [reply ocp_resources/resource.py:153]: To ensure recursion handles lists that are actually list I need to update this code. I was thinking
```
       if isinstance(resource_dict, dict):
        for key, value in resource_dict.items():
            if key == key_name:
                resource_dict[key] = "******"
            elif isinstance(value, dict):
                resource_dict[key] = change_dict_value_to_hashed(value, key_name)
            elif isinstance(value, list):
                for key_list, value_list in enumerate(value):
                    value[key_list] = change_dict_value_to_hashed(value_list, key_name)
    if isinstance(resource_dict, list):
        for key_list, value_list in enumerate(resource_dict):
            if isinstance(value_list, dict):
                resource_dict[key_list] = change_dict_value_to_hashed(value_list, key_name)
    return resource_dict
```

File: ocp_resources/sealed_secret.py (hunk 4 above)
Comment (dbasunag) [reply ocp_resources/sealed_secret.py:53]: ok, so may be if a.1 is passed then only mask a.1 else if "a" is passed, then mask all occurrences of "a"?

File: ocp_resources/resource.py (hunk 1 above)
Comment (dbasunag) [reply ocp_resources/resource.py:136]: Done.

File: ocp_resources/resource.py (hunk 2 above)
Comment (dbasunag) [reply ocp_resources/resource.py:146]: This is using recursion. So raising ValueError may not be a good idea.

Comment (dbasunag) [reply ocp_resources/resource.py:146]: I will do this one. My mistake.

File: ocp_resources/sealed_secret.py (hunk 4 above)
Comment (dbasunag) [reply ocp_resources/sealed_secret.py:53]: Ok, I will update the PR based on our discussion today. Thanks a lot!

File: ocp_resources/resource.py (hunk 2 above)
Comment (dbasunag) [reply ocp_resources/resource.py:147]: Moving this to hash_resource_dict()

File: ocp_resources/resource.py [hunk 6]
```diff
@@ -134,9 +134,38 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+            for resource_element in resource_dict[key_lists[0]]:
+                if isinstance(resource_element, dict):
+                    resource_element = benedict(resource_element, keypath_separator=">")
+                    updated_resource_list.append(
+                        replace_key_with_hashed_value(resource_dict=resource_element, key_name=key_lists[1])
+                    )
+                else:
+                    updated_resource_list.append(resource_element)
+            resource_dict[key_lists[0]][:] = updated_resource_list
+    return resource_dict
```
[the remark this answers is not in the document]
Comment (myakove) [reply ocp_resources/resource.py:168]: Looks better, @dbasunag please check if this works

**Review by myakove — CHANGES_REQUESTED** [review]

File: tests/test_validate_resources.py [hunk 7]
```diff
@@ -18,3 +32,50 @@ def resources_definitions_errors() -> List[str]:
...
+    }
+    result = replace_key_with_hashed_value(resource_dict=data, key_name="a[]>b")
+    assert result["a"][0]["b"] == "*******"
+    assert result["a"][1]["b"] == "*******"
+
+
+def test_replace_key_with_hashed_value_invalid_inputs():
+    """Test that the function handles invalid inputs appropriately."""
+    with pytest.raises(TypeError):
+        replace_key_with_hashed_value(resource_dict=None, key_name="a>b")
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_validate_resources.py:81]: @myakove do you think it would be useful. The existing tests are very similar I feel.

File: ocp_resources/resource.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbasunag) [reply ocp_resources/resource.py]: Done. It did work!

File: ocp_resources/resource.py [hunk 8]
```diff
@@ -134,9 +134,38 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+    Recursively search a nested dictionary for a given key and changes its value to "******" if found.
 
-# End Exceptions classes
+    Args:
+        resource_dict: The nested dictionary to search.
+        key_name: The key to find.
+
+    Returns:
+        The modified dictionary.
+    """
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply ocp_resources/resource.py]: Done!

File: ocp_resources/resource.py [hunk 9]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
     return None
 
 
-# Exceptions classes
+def replace_key_with_hashed_value(resource_dict: Dict[Any, Any], key_name: str) -> Dict[Any, Any]:
+    """
+    Recursively search a nested dictionary for a given key and changes its value to "******" if found.
+    The function supports two key formats:
+    1. Regular path: "path>to>key"
+    2. List path: "path>to[]>key"
```
Comment (myakove) [inline ocp_resources/resource.py:142]: Please fix the docstring.

empty line before Args and Return and Raises

File: ocp_resources/resource.py [hunk 10]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
     return None
 
 
-# Exceptions classes
+def replace_key_with_hashed_value(resource_dict: Dict[Any, Any], key_name: str) -> Dict[Any, Any]:
+    """
+    Recursively search a nested dictionary for a given key and changes its value to "******" if found.
+    The function supports two key formats:
+    1. Regular path: "path>to>key"
```
Comment (myakove) [inline ocp_resources/resource.py:141]: Please add a simple example of a dict and the query that pass to it.

Comment (myakove) [inline ocp_resources/resource.py:141]: You said we have 2 formats, but you didn't explain their differences.

File: ocp_resources/resource.py [hunk 11]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+        return dict(benedict_resource_dict)
+
+    # Handle list case
+    key_prefix, remaining_key = key_name.split("[]>", 1)
+    if not benedict_resource_dict.get(key_prefix):
+        return dict(benedict_resource_dict)
 
-# End Exceptions classes
+    # Process list elements
+    target_list = benedict_resource_dict[key_prefix]
```
Comment (myakove) [inline ocp_resources/resource.py:168]: please rename `target_list` it's not a list as far as we know.

File: ocp_resources/resource.py [hunk 12]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+    Recursively search a nested dictionary for a given key and changes its value to "******" if found.
+    The function supports two key formats:
+    1. Regular path: "path>to>key"
+    2. List path: "path>to[]>key"
+    Args:
+        resource_dict: The nested dictionary to search. Must be a dictionary.
+        key_name: The key path to find. Must be a string in the format "path>to>key" or "path>to[]>key".
+    Returns:
+        Dict[Any, Any]: A copy of the input dictionary with the specified key's value replaced with "*******".
+    Raises:
```
Comment (myakove) [inline ocp_resources/resource.py:148]: this function do not raise anything

File: tests/test_validate_resources.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (myakove) [reply tests/test_validate_resources.py:81]: +1, real world will be better

File: tests/test_validate_resources.py [hunk 13]
```diff
@@ -18,3 +32,50 @@ def resources_definitions_errors() -> List[str]:
...
+def test_replace_key_with_hashed_value_replace_key_in_list(nested_dict):
+    assert replace_key_with_hashed_value(resource_dict=nested_dict, key_name="d[]>b>c") == {
+        "a": {"b": "c"},
+        "d": [
+            {"b": {"c": "*******"}},
+            {
+                "e": {"f": "g"},
+            },
+        ],
+    }
```
Comment (myakove) [inline tests/test_validate_resources.py:58]: should be parametrazied.

File: tests/test_validate_resources.py [hunk 14]
```diff
@@ -2,6 +2,20 @@
 import pytest
 
 from tests.scripts.validate_resources import parse_resource_file_for_errors, resource_file
+from ocp_resources.resource import replace_key_with_hashed_value
```
Comment (myakove) [inline tests/test_validate_resources.py:5]: please add negative tests as well

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/resource.py (hunk 10 above)
Comment (dbasunag) [reply ocp_resources/resource.py:141]: Added examples. Please let me know if this works.

File: ocp_resources/resource.py (hunk 11 above)
Comment (dbasunag) [reply ocp_resources/resource.py:168]: Changed to resource_data

File: ocp_resources/resource.py [hunk 15]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
     return None
 
 
-# Exceptions classes
+def replace_key_with_hashed_value(resource_dict: Dict[Any, Any], key_name: str) -> Dict[Any, Any]:
+    """
+    Recursively search a nested dictionary for a given key and changes its value to "******" if found.
+    The function supports two key formats:
```
Comment (rnetser) [inline ocp_resources/resource.py:140]: missing newline

File: ocp_resources/resource.py [hunk 16]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+    1. Regular path: "path>to>key"
+    2. List path: "path>to[]>key"
+    Args:
+        resource_dict: The nested dictionary to search. Must be a dictionary.
```
Comment (rnetser) [inline ocp_resources/resource.py:144]:
```suggestion
        resource_dict: The nested dictionary to search. 
```
no need for this; the function has typing

File: ocp_resources/resource.py (hunk 10 above)
Comment (rnetser) [reply ocp_resources/resource.py:141]: i do not understand the examples; can you provide real ones? i.e example dict and example path

File: ocp_resources/resource.py [hunk 17]
```diff
@@ -134,9 +134,47 @@ def sub_resource_level(current_class: Any, owner_class: Any, parent_class: Any)
...
+        key_name: The key path to find. Must be a string in the format "path>to>key" or "path>to[]>key".
+    Returns:
+        Dict[Any, Any]: A copy of the input dictionary with the specified key's value replaced with "*******".
+    Raises:
+        ValueError: If resource_dict is not a dictionary or key_name is not a string.
+    """
+    # Create a deep copy to avoid modifying the input
```
Comment (rnetser) [inline ocp_resources/resource.py:151]: please remove the inline comments; the code is clear

File: ocp_resources/resource.py [hunk 18]
```diff
@@ -1171,21 +1209,18 @@ def keys_to_hash(self) -> List[str]:
...
          Example:
              given a dict: {"spec": {"data": <value_to_hash>}}
-             To hash spec['data'] key pass: ["spec..data"]
+             To hash spec['data'] key pass: ["spec>data"]
         """
         return []
 
     def hash_resource_dict(self, resource_dict: Dict[Any, Any]) -> Dict[Any, Any]:
+        if not isinstance(resource_dict, dict):
+            raise ValueError("Expected a dictionary as the first argument")
```
Comment (rnetser) [inline ocp_resources/resource.py:1218]: why is this needed? we do not check types

File: tests/test_validate_resources.py [hunk 19]
```diff
@@ -18,3 +32,50 @@ def resources_definitions_errors() -> List[str]:
...
+    }
+
+
+def test_replace_key_with_hashed_value_edge_cases():
+    """Test edge cases for replace_key_with_hashed_value function."""
+    # Empty dictionary
+    assert replace_key_with_hashed_value(resource_dict={}, key_name="a>b") == {}
+
+    # Non-existent key
+    assert replace_key_with_hashed_value(resource_dict={"x": {"y": "z"}}, key_name="a>b") == {"x": {"y": "z"}}
```
Comment (rnetser) [inline tests/test_validate_resources.py:67]: each should be a separate test

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/resource.py (hunk 10 above)
Comment (dbasunag) [reply ocp_resources/resource.py:141]: I didn't push the patch yet. I will do that.

File: ocp_resources/resource.py (hunk 17 above)
Comment (dbasunag) [reply ocp_resources/resource.py:151]: ok, those were from coderabbit :). Will remove.

File: ocp_resources/resource.py (hunk 18 above)
Comment (dbasunag) [reply ocp_resources/resource.py:1218]: @myakove suggested adding this one.

Comment (myakove) [reply ocp_resources/resource.py:1220]: It's better to have a nice error when failing, typing is not enforced, only suggests the type (by IDE)

File: ocp_resources/resource.py (hunk 15 above)
Comment (rnetser) [reply ocp_resources/resource.py:140]: the newline should come after the oneliner function description

File: ocp_resources/resource.py (hunk 18 above)
Comment (rnetser) [reply ocp_resources/resource.py:1220]: but we do not have such checks in the wrapper

**Review by rnetser — CHANGES_REQUESTED** [review]

File: tests/test_validate_resources.py (hunk 14 above)
Comment (dbasunag) [reply tests/test_validate_resources.py:6]: I have added some of the negative tests.

File: ocp_resources/resource.py (hunk 18 above)
Comment (dbasunag) [reply ocp_resources/resource.py:1220]: Let me know how do you want me to proceed.

Comment (dbasunag) [reply ocp_resources/resource.py:1251]: @myakove / @rnetser please let me know how you want me to proceed here. I will update accordingly.

Comment (myakove) [reply ocp_resources/resource.py:1251]: leave the error with the msg, we should fail nice

File: tests/test_validate_resources.py [hunk 20]
```diff
@@ -1,7 +1,25 @@
 from typing import List
 import pytest
+from benedict import benedict
 
 from tests.scripts.validate_resources import parse_resource_file_for_errors, resource_file
+from ocp_resources.resource import replace_key_with_hashed_value
+
+
+@pytest.fixture()
```
Comment (myakove) [inline tests/test_validate_resources.py:9]: should not be here, create a new file for unittests

**Review by myakove — CHANGES_REQUESTED** [review]

File: tests/test_replace_key_with_hashed_value.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
+import pytest
```
Comment (myakove) [inline tests/test_replace_key_with_hashed_value.py:1]: I was thinking about test_unittests.py file name

**Review by myakove — CHANGES_REQUESTED** [review]

**Comment by myakove** [comment]
And please check coderabbitai comments

File: tests/test_unittests.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    """Test edge cases for replace_key_with_hashed_value function."""
+    assert replace_key_with_hashed_value(resource_dict=resource, key_name=key_name) == expected_result
+
+
+def test_replace_key_with_hashed_value_multiple_occurances(vm_spec):
+    vm_spec["spec"]["template"]["spec"]["volumes"].append({"name": "volume2", "userData": "more sensitive-data"})
+    result = replace_key_with_hashed_value(resource_dict=vm_spec, key_name="spec>template>spec>volumes[]>userData")
+    for volume in result["spec"]["template"]["spec"]["volumes"]:
+        assert volume["userData"] == "*******"
+
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_unittests.py:67]: It is a function scoped fixture. So would not be able to impact other tests. List order should not matter either.

File: tests/test_unittests.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+    ],
+)
+def test_replace_key_with_hashed_value(vm_spec, key_name, expected_key):
+    assert (
+        benedict(replace_key_with_hashed_value(resource_dict=vm_spec, key_name=key_name), keypath_separator=">")[
+            expected_key
+        ]
+        == "*******"
+    )
+
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_unittests.py:44]: This does not seem right, keypaths would be `['spec', 'spec>template', 'spec>template>features', 'spec>template>features>someNewFeature', 'spec>template>features>someNewFeature>someNewData', 'spec>template>spec', 'spec>template>spec>volumes']`, they would never match `spec>template>spec>volumes[0]>userData`.  I can remove expected_key from both the dicts and compare that they are same.

File: tests/test_unittests.py [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    vm_spec["spec"]["template"]["spec"]["volumes"].append({"name": "volume2", "userData": "more sensitive-data"})
+    result = replace_key_with_hashed_value(resource_dict=vm_spec, key_name="spec>template>spec>volumes[]>userData")
+    for volume in result["spec"]["template"]["spec"]["volumes"]:
+        assert volume["userData"] == "*******"
+
+
+def test_replace_key_with_hashed_value_invalid_inputs():
+    """Test that the function handles invalid inputs appropriately."""
+    with pytest.raises(TypeError):
+        replace_key_with_hashed_value(resource_dict=None, key_name="a>b")
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_unittests.py:72]:
```with pytest.raises(TypeError):
        replace_key_with_hashed_value(resource_dict=[], key_name="a>b")
``` 
does not seem valid. Adding the rests.

File: tests/test_unittests.py [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+        # Non-existent key
+        pytest.param(
+            {"x": {"y": "z"}}, "a>b", {"x": {"y": "z"}}, id="test_replace_key_with_hashed_value_key_doesnt_exist"
+        ),
+    ],
+)
+def test_replace_key_with_hashed_value_edge_cases(resource, key_name, expected_result):
+    """Test edge cases for replace_key_with_hashed_value function."""
+    assert replace_key_with_hashed_value(resource_dict=resource, key_name=key_name) == expected_result
+
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_unittests.py:60]: Added the empty key path and malformed key path. deep nesting test already exists.

File: tests/test_replace_key_with_hashed_value.py (hunk 21 above)
Comment (dbasunag) [reply tests/test_replace_key_with_hashed_value.py]: Done.

File: tests/test_replace_key_with_hashed_value.py [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+    vm_spec["spec"]["template"]["spec"]["volumes"].append({"name": "volume2", "userData": "more sensitive-data"})
+    result = replace_key_with_hashed_value(resource_dict=vm_spec, key_name="spec>template>spec>volumes[]>userData")
+    for volume in result["spec"]["template"]["spec"]["volumes"]:
+        assert volume["userData"] == "*******"
+
+
+def test_replace_key_with_hashed_value_invalid_inputs():
+    """Test that the function handles invalid inputs appropriately."""
+    with pytest.raises(TypeError):
+        replace_key_with_hashed_value(resource_dict=None, key_name="a>b")
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_replace_key_with_hashed_value.py]: Added these tests.

**Review by myakove — APPROVED** [review]
